lib/AST/LocalizationFormat.cpp

```diff
@@ -143,30 +137,48 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+llvm::Optional<uint32_t> getDiagnosticID(llvm::StringRef rawID) {
+  llvm::Optional<uint32_t> id;
+  auto it = llvm::find_if(diagnosticID,
+                          [&](const char *diagID) { return rawID == diagID; });
```

Comment:
Is there any way we can avoid linear scan here? This means that parsing YAML is going to be quadratic which is really unfortunate...

Reply:
I think I can make `diagnosticID` a `std::map` or `std::unordered_map` something like the following:
```cpp
static const std::map<std::string, LocalDiagID> diagnosticID = {
#define DIAG(KIND, ID, Options, Signature)\
{#ID, ID},
#include "...\DiagnosticsAll.def"
};
```
That way if I'm looking up to see if an ID is in the map it'll be a constant time or logarithmic time, depending on `std::map` or `std::unordered_map`. WDYT? Also if you know any llvm equivalent data structure, please suggest it. :)

Reply:
Yeah, we could use a `std::map` at the end but I'd rather explore a possibility I have outlined regarding `yamlize`.